test/integration/resume_request_test.go

```diff
@@ -0,0 +1,183 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package integration
+
+import (
+	"context"
+	"crypto"
+	"crypto/x509"
+	"os"
+	"reflect"
+	"testing"
+	"time"
+
+	cmclient "github.com/cert-manager/cert-manager/pkg/client/clientset/versioned"
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	fakeclock "k8s.io/utils/clock/testing"
+
+	"github.com/cert-manager/csi-lib/manager"
+	"github.com/cert-manager/csi-lib/metadata"
+	"github.com/cert-manager/csi-lib/storage"
+	testdriver "github.com/cert-manager/csi-lib/test/driver"
+	testutil "github.com/cert-manager/csi-lib/test/util"
+)
+
+func testResumesExistingRequest(t *testing.T, issueBeforeCall bool) {
+	store := storage.NewMemoryFS()
+	ns := "certificaterequest-namespace"
+	clock := fakeclock.NewFakeClock(time.Now())
+	opts, cl, stop := testdriver.Run(t, testdriver.Options{
+		Store: store,
+		Clock: clock,
+		GeneratePrivateKey: func(meta metadata.Metadata) (crypto.PrivateKey, error) {
+			return nil, nil
+		},
+		GenerateRequest: func(meta metadata.Metadata) (*manager.CertificateRequestBundle, error) {
+			return &manager.CertificateRequestBundle{
+				Namespace: ns,
+			}, nil
+		},
+		SignRequest: func(meta metadata.Metadata, key crypto.PrivateKey, request *x509.CertificateRequest) (csr []byte, err error) {
+			return []byte{}, nil
+		},
+		WriteKeypair: func(meta metadata.Metadata, key crypto.PrivateKey, chain []byte, ca []byte) error {
+			store.WriteFiles(meta, map[string][]byte{
+				"ca":   ca,
+				"cert": chain,
+			})
+			nextIssuanceTime := clock.Now().Add(time.Hour)
+			meta.NextIssuanceTime = &nextIssuanceTime
+			return store.WriteMetadata(meta.VolumeID, meta)
+		},
+	})
+	defer stop()
+
+	tmpDir, err := os.MkdirTemp("", "*")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.RemoveAll(tmpDir)
+
+	// create a root, non-expiring context
+	ctx := context.Background()
+
+	// We are going to submit this request multiple times, so lets just write it out once
+	nodePublishVolumeRequest := &csi.NodePublishVolumeRequest{
+		VolumeId: "test-vol",
+		VolumeContext: map[string]string{
+			"csi.storage.k8s.io/ephemeral":     "true",
+			"csi.storage.k8s.io/pod.name":      "the-pod-name",
+			"csi.storage.k8s.io/pod.namespace": ns,
+		},
+		TargetPath: tmpDir,
+		Readonly:   true,
+	}
+
+	// create a context that expires in 2s (enough time for at least a single call of `issue`)
+	twoSecondCtx, cancel := context.WithTimeout(ctx, time.Second*2)
+	defer cancel()
```

Comment:
Consider using `t.Cleanup` instead of `defer` for contexts too.
```suggestion
	t.Cleanup(cancel)
```